### PR TITLE
feat: idx interaction code flow integration

### PR DESCRIPTION
**Pull request by shuowu**
## PR Checklist

- [ ] Have you verified the basic functionality for this change?

- [ ] Added unit tests?

- [ ] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Issue:

- [OKTA-XXXXXX](https://oktainc.atlassian.net/browse/OKTA-XXXXXX)

### CODE REVIEW

File: src/v2/BaseLoginRouter.js [hunk 1]
```diff
@@ -32,11 +32,21 @@ import AppState from './models/AppState';
...
+const introspect = (settings) => {
+  const clientId = settings.get('clientId');
   const domain = settings.get('baseUrl');
+  const scopes = settings.get('scopes');
   const stateHandle = settings.get('stateToken');
+  const redirectUri = settings.get('redirectUri');
   const version = settings.get('apiVersion');
-  return idx.start({ domain, stateHandle, version });
+  
+  if (settings.get('useInteractionCodeFlow')) {
```
Comment (swiftone) [inline src/v2/BaseLoginRouter.js:43]: I would check stateHandle first - A customized-okta-hosted widget could be configured with this switch, but we would want to maintain the okta-hosted flow.

File: src/v2/BaseLoginRouter.js [hunk 2]
```diff
@@ -32,11 +32,21 @@ import AppState from './models/AppState';
 import idx from 'idx';
 
 
-const introspectStateToken = (settings) => {
+const introspect = (settings) => {
```
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:35]: I would like to maintain the separation between introspect and interact calls..

create two functions introspect and interact that returns idx.start()

create another calling function startLoginFlow() that calls the correct function from above

File: src/v2/BaseLoginRouter.js [hunk 3]
```diff
@@ -32,11 +32,21 @@ import AppState from './models/AppState';
...
+  const clientId = settings.get('clientId');
   const domain = settings.get('baseUrl');
+  const scopes = settings.get('scopes');
   const stateHandle = settings.get('stateToken');
+  const redirectUri = settings.get('redirectUri');
   const version = settings.get('apiVersion');
-  return idx.start({ domain, stateHandle, version });
+  
+  if (stateHandle) {
```
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:43]: what if `stateHandle` and `useInteractionCodeFlow` both are present ? In that case we always call `introspect` ?

File: src/v2/BaseLoginRouter.js [hunk 4]
```diff
@@ -32,11 +32,21 @@ import AppState from './models/AppState';
...
+    return idx.start({ domain, stateHandle, version });
+  } else if (settings.get('useInteractionCodeFlow')) {
+    return idx.start({ clientId, domain, scopes, stateHandle, redirectUri, version });
+  } else {
+    throw new Errors.ConfigError('Set "useInteractionCodeFlow" to true in configuration to enable the "interaction_code" flow.');
```
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:48]: need a better to prompt for this warning.. This should not be an error!!

File: src/v2/BaseLoginRouter.js [hunk 5]
```diff
@@ -162,9 +179,11 @@ export default Router.extend({
     // introspect stateToken when widget is bootstrap with state token
     // and remove it from `settings` afterwards as IDX response always has
     // state token (which will be set into AppState)
-    if (this.settings.get('stateToken') || this.settings.get('proxyIdxResponse')) {
+    if (this.settings.get('stateToken') 
+        || this.settings.get('proxyIdxResponse')
+        || this.settings.get('useInteractionCodeFlow')) {
       const idxRespPromise = this.settings.get('proxyIdxResponse') ?
-        handleProxyIdxResponse(this.settings) : introspectStateToken(this.settings);
+        handleProxyIdxResponse(this.settings) : introspect(this.settings);
```
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:186]: call `startLoginFlow` here instead of `introspect`

File: src/widget/OktaSignIn.js [hunk 6]
```diff
@@ -153,7 +153,9 @@ var OktaSignIn = (function () {
     var authClient = createAuthClient(authParams);
 
     var Router;
-    if ((options.stateToken && !Util.isV1StateToken(options.stateToken)) || options.proxyIdxResponse) {
+    if ((options.stateToken && !Util.isV1StateToken(options.stateToken)) 
+        || options.useInteractionCodeFlow
```
Comment (pradeepdewda-okta) [inline src/widget/OktaSignIn.js:157]: at this moment we don't know if IDX flag is enabled or not... I think if stateToken exists, then only we can se IDX flag was enabled and hence we can use `useInteractionCodeFlow`..

File: src/v2/BaseLoginRouter.js (hunk 4 above)
Comment (shuowu) [reply src/v2/BaseLoginRouter.js:48]: The error here is for devs only, I assume the above two cases are the only cases we should support.

File: src/v2/BaseLoginRouter.js [hunk 7]
```diff
@@ -32,11 +32,21 @@ import AppState from './models/AppState';
...
+  const scopes = settings.get('scopes');
   const stateHandle = settings.get('stateToken');
+  const redirectUri = settings.get('redirectUri');
   const version = settings.get('apiVersion');
-  return idx.start({ domain, stateHandle, version });
+  
+  if (stateHandle) {
+    return idx.start({ domain, stateHandle, version });
+  } else if (settings.get('useInteractionCodeFlow')) {
+    return idx.start({ clientId, domain, scopes, stateHandle, redirectUri, version });
```
Comment (swiftone) [inline src/v2/BaseLoginRouter.js:46]: This should be issuer, not domain, if we have issuer.  (the stateHandle version can continue to use domain, or can also switch over to issuer)

File: src/widget/OktaSignIn.js (hunk 6 above)
Comment (shuowu) [reply src/widget/OktaSignIn.js:157]: The `useInteractionCodeFlow` runs the interaction code flow, which boostraps without stateToken. Should the logic be if any of the conditions satisfied, then use V2Router to use idx?

File: src/v2/BaseLoginRouter.js (hunk 3 above)
Comment (shuowu) [reply src/v2/BaseLoginRouter.js:43]: I think Brett's comment can answer your question: https://github.com/okta/okta-signin-widget/pull/1533#discussion_r530637890

File: src/v2/BaseLoginRouter.js (hunk 2 above)
Comment (swiftone) [reply src/v2/BaseLoginRouter.js:35]: interact is called inside idx.start(), so the widget doesn't explicitly call it.  If you mean two function calls to define the different params to idx.start(), then ignore this comment from me.

File: src/widget/OktaSignIn.js (hunk 6 above)
Comment (pradeepdewda-okta) [reply src/widget/OktaSignIn.js:157]: This issue is we need to check that IDX flag was enabled or not.. Till now we expected that if `stateToken` exists then IDX flag is enabled by default. But now it seems it's not the case.

File: src/v2/BaseLoginRouter.js (hunk 3 above)
Comment (swiftone) [reply src/v2/BaseLoginRouter.js:43]: If both are present, we should behave as if just stateHandle is present and ignore useInteractionCodeFlow.   This should only happen if there is a customized-okta-hosted widget, and in such a case we want to end with a redirect, so ignore useInteractionCodeFlow.

Comment (pradeepdewda-okta) [reply src/v2/BaseLoginRouter.js:43]: @bretterer do we always call `introspect` with both exists ?

File: src/widget/OktaSignIn.js (hunk 6 above)
Comment (shuowu) [reply src/widget/OktaSignIn.js:157]: like a feature flag for `idx`? if there is one, only that flag should be enough here.

File: src/v2/BaseLoginRouter.js (hunk 2 above)
Comment (pradeepdewda-okta) [reply src/v2/BaseLoginRouter.js:35]: yes, for the purpose of clarity for devs.

File: test/unit/spec/OktaSignIn_spec.js [hunk 8]
```diff
@@ -402,9 +401,65 @@ Expect.describe('OktaSignIn v2 bootstrap', function () {
...
+    itp('throws an error if invalid version is passed to idx-js', function () {
+      return setupLoginFlow({
         apiVersion: '2.0.0',
-      }).catch(err => {
+        clientId: 'someClientId',
+        redirectUri: 'http://0.0.0.0:9999',
+        useInteractionCodeFlow: true
+      }, responses).catch(err => {
         expect(err.name).toBe('CONFIG_ERROR');
         expect(err.message.toString()).toEqual('Error: Unknown api version: 2.0.0.  Use an exact semver version.');
```
Comment (pradeepdewda-okta) [inline test/unit/spec/OktaSignIn_spec.js:464]: can we get the exact api version number here which is being used ? instead of hardcoding

**Review by swiftone — APPROVED** [review]

File: src/v2/BaseLoginRouter.js [hunk 9]
```diff
@@ -160,18 +210,20 @@ export default Router.extend({
...
-    if (this.settings.get('stateToken') || this.settings.get('proxyIdxResponse')) {
+    if (this.settings.get('oieEnabled')) {
       const idxRespPromise = this.settings.get('proxyIdxResponse') ?
-        handleProxyIdxResponse(this.settings) : introspectStateToken(this.settings);
+        handleProxyIdxResponse(this.settings) : startLoginFlow(this.settings);
       return idxRespPromise
         .then(idxResp => {
           this.settings.unset('stateToken');
           this.settings.unset('proxyIdxResponse');
+          this.settings.unset('useInteractionCodeFlow');
```
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:220]: Can you please create a new `unsetAll` method as well, that can take an array and unset all..

File: src/v2/BaseLoginRouter.js [hunk 10]
```diff
@@ -160,18 +210,20 @@ export default Router.extend({
     // introspect stateToken when widget is bootstrap with state token
     // and remove it from `settings` afterwards as IDX response always has
     // state token (which will be set into AppState)
-    if (this.settings.get('stateToken') || this.settings.get('proxyIdxResponse')) {
+    if (this.settings.get('oieEnabled')) {
```
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:213]: thanks for this.. 🥇

**Pull request closed without merging**